### PR TITLE
🩹 Date Inconsistencies – Events Component

### PR DESCRIPTION
**Pull request by isaaclombardssw**
This was a real tricky one – as per PBI #2129 the date convertions meant that event dates weren't totally reliable.

3 issues....

(1) Using tina to enter an event date would **assume your local time is being used and convert it to UTC.**
==> saved value = Actual time- offset of content editor vs UTC.
Actual time being the event time in the event timezone.
WANT only actual time.

(2) The date appearing in the UI is cast to a string via casting functions that have a hidden conversion into local time. We want what's shown in the UI to be accurate with respect to the event location.
==> shown value = (Actual time - offset vs UTC of content editor) + offset vs UTC of site user.
WANT only actual time.

(3) The live indicator was dependent on this saved value, so for example an event would be live if...
==> (Actual time - offset vs UTC of content editor) + offset vs UTC of site user > Users current time in local timezone
WANT actual time >  users current time - offset vs UTC of site user + offset vs UTC of event timezone.

This has been solved by allowing...

1. The event time to be specified.
2. The event date is now entered and stored in the event timezone (in practice it's all stored/entered in "UTC" to prevent any conversions Tina is doing with the Date object, with offsets specified manually to fit the above equations).
3. The offset vs UTC of the event timezone to be specified.

This required a couple of workarounds based on how the Tina date field operates. 

- Setting the date picker to UTC is possible but decouples it from the associated Tina editor input box, which still tries to cast it into local time. A format has been introduced to fix that.
- Due to the format, the time picker appears to be bricked, so a separate field was introduced to specify the time (typically Tina could handle both at once), separate from the date. This has been validated to keep it between 0 and 23 (24 hour time) integers.
- As 'utc', 'step' and some other fields was causing linting errors despite working and being included in the docs, I've had to add a ts-ignore to let the type pass... 
utc prop can be used as per the [date component](https://tina.io/docs/reference/toolkit/fields/date/#datetimepickerprops) docs, and step is something that exist on the [number component](https://tina.io/docs/reference/toolkit/fields/number/) spec. 

<details>

### General Contributing:

- [ ] Have you followed the guidelines in our [Contributing document](https://github.com/tinacms/tina.io/blob/master/CONTRIBUTING.md)?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### All New Content Submissions: (To be confirmed by reviewer)

- [ ] Title is short & specific
- [ ] Headers are logically ordered & consistent
- [ ] Purpose of document is explained in the first paragraph
- [ ] Procedures are tested and work
- [ ] Any technical concepts are explained or linked to
- [ ] Document follows structure from templates
- [ ] All links work
- [ ] The spelling and grammar checker has been run
- [ ] Graphics and images are clear and useful
- [ ] Any prerequisites and next steps are defined.

</details>

### CODE REVIEW

File: components/blocks/Events.template.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+      label: 'Card Items',
+      type: 'object',
+      list: true,
+      ui: {
+        itemProps: (item) => ({
+          key: item.id,
+          label: item.headline,
+        }),
+      },
+      // @ts-ignore: type error as utc, options and step fields aren't formally recognised but valid as per docs
```
Comment (joshbermanssw) [inline components/blocks/Events.template.tsx:49]: link to docs pls

File: components/blocks/Events.template.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            utc: true,
+            format: (value, name, field) => value && timeFormat.format(new Date(Date.parse(value)))
+          }, 
+        },
+        {
+          name: 'startTime',
+          label: 'Start Time (24 hour time)',
+          type: 'number',
+          description:
+            'Optional hours field for more accurate "Live"/"Done" chips on the event card. 24 hours time, ex. 14 = 2:00pm',
```
Comment (joshbermanssw) [inline components/blocks/Events.template.tsx:68]: a bit long and repeating the 24 hour stuff i think we should make it mandatory

File: components/blocks/Events.template.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            utc: true,
+            format: (value, name, field) => value && timeFormat.format(new Date(Date.parse(value)))
+          }, 
+        },
+        {
+          name: 'endTime',
+          label: 'End Time (24 hour time)',
+          type: 'number',
+          description:
+            'Optional field for more accurate "Live"/"Done" chips on the event card. 24 hours time, ex. 14 = 2:00pm.',
```
Comment (joshbermanssw) [inline components/blocks/Events.template.tsx:90]: same as above (i.e end of event time)

**Review by joshbermanssw — COMMENTED** [review]
few smaller template changes

File: components/blocks/Events.template.tsx (hunk 1 above)
Comment (isaaclombardssw) [reply components/blocks/Events.template.tsx:49]: Done.

File: components/blocks/Events.template.tsx (hunk 2 above)
Comment (isaaclombardssw) [reply components/blocks/Events.template.tsx:68]: Chatting to @landonmaxwell and Gert we've decided to make the start time mandatory and remove the end time entirely.

**Comment by isaaclombardssw** [comment]
As per conversation with @landonmaxwell and @Marxoz  – end time isn't necessary so has been removed / defaults to midnight.

File: components/blocks/Events.template.tsx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  if (value && value % 1 != 0) {
+    return "Only whole numbers should be used."
+  }
+}
+
+export const eventsTemplate: Template = {
+  label: 'Events',
+  name: 'events',
+  ui: {
+    previewSrc: '',
```
Comment (bradystroud) [inline components/blocks/Events.template.tsx:34]: Can we add a screenshot here?

File: components/blocks/Events.tsx [hunk 5]
```diff
@@ -21,66 +21,32 @@ const Card = ({ cardItem, onHover }) => {
...
   }
 
-  const daysUntilEvent = calculateDaysUntilEvent(cardItem.startDate)
-  const isPastEvent = daysUntilEvent < 0
+  const startDateUTC = new Date(Date.parse(cardItem.startDate))
+  startDateUTC.setUTCMinutes(startDateUTC.getUTCMinutes() + (cardItem.timezone * -60) + ((cardItem.startTime) * 60))
+  const endDateUTC = new Date(Date.parse(cardItem.endDate ?? cardItem.startDate))
+  endDateUTC.setUTCMinutes(endDateUTC.getUTCMinutes() + (cardItem.timezone * -60) + (24 * 60))
+  const hoursUntilEvent = Math.ceil((startDateUTC.getTime() - (new Date()).getTime()) / 36e5)
+  const hoursUntilEventEnd = Math.ceil((endDateUTC.getTime() - (new Date()).getTime()) / 36e5)
```
Comment (bradystroud) [inline components/blocks/Events.tsx:44]: this is tricky to read
Can we add some comments to help explain it?

File: components/blocks/Events.template.tsx [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+          name: 'startTime',
+          label: 'Start Time',
+          type: 'number',
+          description:
+            'Hours field for more accurate "Live"/"Done" chips on the event card. 24 hours time, ex. 14 = 2:00pm',
+          ui: {
+            step: 1,
+            validate: timezoneValidation
+          },
+        },
```
Comment (bradystroud) [inline components/blocks/Events.template.tsx:74]: I dont think we need this field, it just adds complexity and the accuracy has not been a problem.

Removing this field means from 12am on the event day, its listed as live. I think this is fine.
Maybe we can rename 'Live' to 'Live today'

**Review by bradystroud — CHANGES_REQUESTED** [review]

File: components/blocks/Events.template.tsx (hunk 6 above)
Comment (isaaclombardssw) [reply components/blocks/Events.template.tsx:74]: "Live today" wouldn't work cause that would be depending on the local time-zones, and I'd need to check what local day corresponds to the event start day in the target timezone.

**Comment by isaaclombardssw** [comment]
As per a chat with Brady – an disabled end time field was added into the UI.

File: components/blocks/Events.tsx (hunk 5 above)
Comment (isaaclombardssw) [reply components/blocks/Events.tsx]: Done ✅

File: components/blocks/Events.template.tsx (hunk 4 above)
Comment (isaaclombardssw) [reply components/blocks/Events.template.tsx]: Done ✅

**Review by bradystroud — APPROVED** [review]
done over the shoulder :)

**Comment by isaaclombardssw** [comment]
Ongoing – some proposed improvements... https://github.com/tinacms/tina.io/issues/2173

**Pull request merged**